### ISSUE TITLE
File::Find not performing as documented

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7913]
Migrated from [rt.perl.org#35847](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=35847) (status was 'resolved')

Searchable as RT35847$

**Comment by p5pRT** [comment]
### From jms@mathras.comcast.net

#### Created by jms@mathras.comcast.net


This is a bug report for perl from jms@&#8203;mathras\.comcast\.net\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.3\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

The use of lstat\(\) is not guaranteed in File&#8203;::Find\, contrary
to its documentation\.

>> FAQ 3\.4&#8203;: How do I find which modules are installed on my system?
>
> It shouldn't matter\.  From "perldoc File&#8203;::Find" \(v\. 1\.07\)&#8203;:
>
>       \*     It is guaranteed that an lstat has been called before the
>             user's "wanted\(\)" function is called\. This enables fast file
>             checks involving  \_\.


Not in all cases\.  lstat\(\) does not always occur in directories
that don't have any subdirectories\.

linux% cd ~/bin  \# Do not run test with \. = dir\-with\-too\-many\-files
linux% cat \.\./temp\.pl
use File&#8203;::Find;
$File&#8203;::Find&#8203;::dont\_use\_nlink = $ARGV\[0\];
my @&#8203;files;
find sub \{ push @&#8203;files\, $File&#8203;::Find&#8203;::name if \-f \_ && /\\\.pm$/ \}\, @&#8203;INC;
print join "\\n"\, @&#8203;files\,'';
linux% perl \.\./temp\.pl 0 | wc \-l
&nbsp;    434
linux% perl \.\./temp\.pl 1 | wc \-l
&nbsp;    943
linux% perl \-v
This is perl\, v5\.8\.3 built for i586\-linux 



linux% diff \-u Find\.pm\.orig Find\.pm
<details><summary>Inline Patch</summary>

```diff
--- Find.pm.orig        2004-02-27 08:31:34.000000000 -0800
+++ Find.pm     2005-05-17 02:39:04.000000000 -0700
@@ -120,8 +120,11 @@
 
 =item *
 
-It is guaranteed that an I<lstat> has been called before the user's
-C<wanted()> function is called. This enables fast file checks involving S< _>.
+Previous versions of File::Find were guaranteed to call an I<lstat>
+before the user's C<wanted()> function was called, but this is no
+longer the case.  Since this depends on File::Find::done_use_nlink, $^O,
+and other factors, fast file checks involving S< _> are not recommended
+unless C<wanted()> calls I<lstat> first.
 
 =item *
 
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.3:

Configured by jms at Tue Feb 17 02:18:23 PST 2004.

Summary of my perl5 (revision 5.0 version 8 subversion 3) configuration:
  Platform:
    osname=linux, osvers=2.4.20-28.9, archname=i586-linux
    uname='linux mathras 2.4.20-28.9 #1 thu dec 18 13:46:46 est 2003 i586 i586 i386 gnulinux '
    config_args='-der'
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O3',
    cppflags='-fno-strict-aliasing -I/usr/include -I/usr/include/gdbm -fno-strict-aliasing -I/usr/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.2 20030222 (Red Hat Linux 3.2.2-5)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/lib'
    libpth=/usr/lib /lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/lib'

Locally applied patches:
    


@INC for perl v5.8.3:
    /usr/lib/perl5/5.8.3/i586-linux
    /usr/lib/perl5/5.8.3
    /usr/lib/perl5/site_perl/5.8.3/i586-linux
    /usr/lib/perl5/site_perl/5.8.3
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.3:
    HOME=/home/jms
    LANG=en_US
    LANGUAGE (unset)
    LANGVAR=en_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/bin:/usr/sbin:/bin:/sbin:/usr/local/bin:/usr/local/sbin:/usr/X11R6/bin:/home/jms/bin
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @schwern

\[jms@&#8203;mathras\.comcast\.net \- Tue May 17 03&#8203;:40&#8203;:07 2005\]&#8203;:
> Not in all cases\.  lstat\(\) does not always occur in directories
> that don't have any subdirectories\.
> 
> linux% cd ~/bin  \# Do not run test with \. = dir\-with\-too\-many\-files
> linux% cat \.\./temp\.pl
> use File&#8203;::Find;
> $File&#8203;::Find&#8203;::dont\_use\_nlink = $ARGV\[0\];
> my @&#8203;files;
> find sub \{ push @&#8203;files\, $File&#8203;::Find&#8203;::name if \-f \_ && /\\\.pm$/ \}\, @&#8203;INC;
> print join "\\n"\, @&#8203;files\,'';
> linux% perl \.\./temp\.pl 0 | wc \-l
>     434
> linux% perl \.\./temp\.pl 1 | wc \-l
>     943
> linux% perl \-v
> This is perl\, v5\.8\.3 built for i586\-linux

I am unable to replicate this problem with either 5\.8\.1\, 5\.8\.6 or
bleadperl@&#8203;24148\.  I don't have a 5\.8\.3 handy to try\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Jul 08\, 2005 at 10&#8203;:10&#8203;:56AM \-0000\, Michael G Schwern via RT wrote&#8203;:
> I am unable to replicate this problem with either 5\.8\.1\, 5\.8\.6 or
> bleadperl@&#8203;24148\.  I don't have a 5\.8\.3 handy to try\.

I can&#8203;:

&nbsp;    \#\!/usr/bin/perl
&nbsp;    mkdir 'x'\, 0777;
&nbsp;    open F\, '>x/f1';
&nbsp;    open F\, '>x/f2';
&nbsp;    open F\, '>x/f3';
&nbsp;    use File&#8203;::Find;
&nbsp;    $File&#8203;::Find&#8203;::dont\_use\_nlink = $ARGV\[0\];
&nbsp;    find sub \{ printf "%s %s\\n"\, \-f \_ ? 'file' : 'notf'\, $File&#8203;::Find&#8203;::name \}\, 'x';
&nbsp;    system 'rm \-r x';

which gives&#8203;:

&nbsp;    $ \./perl \-Ilib /tmp/p1 0
&nbsp;    notf x
&nbsp;    notf x/f2
&nbsp;    notf x/f3
&nbsp;    notf x/f1
&nbsp;    $ \./perl \-Ilib /tmp/p1 1
&nbsp;    notf x
&nbsp;    file x/f2
&nbsp;    file x/f3
&nbsp;    file x/f1

Its a problem as far back as 5\.003\_22 at least\.

Basically\, when it does the 'nlink check' shortcut on a directory to
determine whether the dir only contains files and no subdirs\, it doesn't
bother lstating the individual entries in the dir to determine whether
they're a file or a subdir\.

So we either fix the docs\, as suggested\, or fix the code to stat \*every\*
entry before calling the wanted function\. The latter would defeat the
purpose of the optimisation\, so I vote for the former\.

\-\- 
"But Sidley Park is already a picture\, and a most amiable picture too\.
The slopes are green and gentle\. The trees are companionably grouped at
intervals that show them to advantage\. The rill is a serpentine ribbon
unwound from the lake peaceably contained by meadows on which the right
amount of sheep are tastefully arranged\." \-\- Lady Croom \- Arcadia

**Comment by p5pRT** [comment]
### From @schwern

I still can't reproduce this on OS X\.


On Sun\, Jul 10\, 2005 at 07&#8203;:31&#8203;:02PM \+0100\, Dave Mitchell wrote&#8203;:
> So we either fix the docs\, as suggested\, or fix the code to stat \*every\*
> entry before calling the wanted function\. The latter would defeat the
> purpose of the optimisation\, so I vote for the former\.

I'd be interested to see\, given how much work File&#8203;::Find does anyway\,
just how much a performance hit fixing this would be\.


\-\- 
Michael G Schwern     schwern@&#8203;pobox\.com     http&#8203;://www\.pobox\.com/~schwern
Ahh email\, my old friend\.  Do you know that revenge is a dish that is best 
served cold?  And it is very cold on the Internet\!

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Jul 11\, 2005 at 12&#8203;:35&#8203;:36AM \-0700\, Michael G Schwern wrote&#8203;:
> I still can't reproduce this on OS X\.

I guess the filesystem you have doesn't have the nlinks > 2 property

> On Sun\, Jul 10\, 2005 at 07&#8203;:31&#8203;:02PM \+0100\, Dave Mitchell wrote&#8203;:
> > So we either fix the docs\, as suggested\, or fix the code to stat \*every\*
> > entry before calling the wanted function\. The latter would defeat the
> > purpose of the optimisation\, so I vote for the former\.
> 
> I'd be interested to see\, given how much work File&#8203;::Find does anyway\,
> just how much a performance hit fixing this would be\.

It makes a huge difference on my laptop \(slow disk\)&#8203;:

&nbsp;    use File&#8203;::Find;
&nbsp;    $File&#8203;::Find&#8203;::dont\_use\_nlink = $ARGV\[0\];

&nbsp;    my $count=0;
&nbsp;    find\(sub \{ $count\+\+ \}\, '/usr'\);
&nbsp;    print "count=$count\\n";

running this the first time with arg0 = 0 took about 3 minutes; rerunning
it takes about 10 secs because all the directory reads are cached by the
OS\.

Susequently running with arg0 = 1 \(bearing in mind that the directories
are still cached\) takes about 6 minutes on both the first and subsequent
runs\, presumably because my laptop hasn't got enough free ram to cache
all 300K inodes that have to be read under /usr\.

CPU usage is almost zero; the code is completely IO bound\.

Perhaps the best approrach would be to document that an lstat is no longer
guaranteed\, but add a new option to the find\(\) options hash\, 'lstat'\,
that if true\, reinstates the guarantee\.


\-\- 
Now is the discount of our winter tent
&nbsp;    \-\- sign seen outside camping shop

**Comment by p5pRT** [comment]
### From @schwern

On Sun\, Jul 17\, 2005 at 06&#8203;:42&#8203;:40PM \+0100\, Dave Mitchell wrote&#8203;:
> Perhaps the best approrach would be to document that an lstat is no longer
> guaranteed\, but add a new option to the find\(\) options hash\, 'lstat'\,
> that if true\, reinstates the guarantee\.

Sounds good to me\.  Seems like a waste to guarantee it if you're not going
to use it anyway\.


\-\- 
Michael G Schwern     schwern@&#8203;pobox\.com     http&#8203;://www\.pobox\.com/~schwern
Reality is that which\, when you stop believing in it\, doesn't go away\.
&nbsp;    \-\- Phillip K\. Dick

**Comment by p5pRT** [comment]
### From @rgs

jms@&#8203;mathras\.comcast\.net \(via RT\) wrote&#8203;:
> The use of lstat\(\) is not guaranteed in File&#8203;::Find\, contrary
> to its documentation\.
> 
> >> FAQ 3\.4&#8203;: How do I find which modules are installed on my system?
> >
> > It shouldn't matter\.  From "perldoc File&#8203;::Find" \(v\. 1\.07\)&#8203;:
> >
> >       \*     It is guaranteed that an lstat has been called before the
> >             user's "wanted\(\)" function is called\. This enables fast file
> >             checks involving  \_\.
> 
> 
> Not in all cases\.  lstat\(\) does not always occur in directories
> that don't have any subdirectories\.
\.\.\.
> \-\-\- Find\.pm\.orig        2004\-02\-27 08&#8203;:31&#8203;:34\.000000000 \-0800
> \+\+\+ Find\.pm     2005\-05\-17 02&#8203;:39&#8203;:04\.000000000 \-0700

Thanks\, doc patch applied as change \#26076 to bleadperl\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rgs

Rafael Garcia\-Suarez wrote&#8203;:
> jms@&#8203;mathras\.comcast\.net \(via RT\) wrote&#8203;:
> > The use of lstat\(\) is not guaranteed in File&#8203;::Find\, contrary
> > to its documentation\.
> > 
> > >> FAQ 3\.4&#8203;: How do I find which modules are installed on my system?
> > >
> > > It shouldn't matter\.  From "perldoc File&#8203;::Find" \(v\. 1\.07\)&#8203;:
> > >
> > >       \*     It is guaranteed that an lstat has been called before the
> > >             user's "wanted\(\)" function is called\. This enables fast file
> > >             checks involving  \_\.
> > 
> > 
> > Not in all cases\.  lstat\(\) does not always occur in directories
> > that don't have any subdirectories\.
> \.\.\.
> > \-\-\- Find\.pm\.orig        2004\-02\-27 08&#8203;:31&#8203;:34\.000000000 \-0800
> > \+\+\+ Find\.pm     2005\-05\-17 02&#8203;:39&#8203;:04\.000000000 \-0700
> 
> Thanks\, doc patch applied as change \#26076 to bleadperl\.

As this documentation appears on a section about follow and follow\_fast\,
I tweaked it further as follows : \(cc&#8203;:ing brian because of the perlfaq
change\.\)

Change 26128 by rgs@&#8203;bloom on 2005/11/14 15&#8203;:40&#8203;:08

&nbsp;        A better fix for \[perl \#35847\] File&#8203;::Find not performing as documented\,
&nbsp;        suggested by Darren Dunham\. Includes a fix to the code example that
&nbsp;        uses File&#8203;::Find in perlfaq3\.

Affected files \.\.\.

\.\.\. //depot/perl/lib/File/Find\.pm\#83 edit
\.\.\. //depot/perl/pod/perlfaq3\.pod\#89 edit

Differences \.\.\.

==== //depot/perl/lib/File/Find\.pm\#83 \(text\) ====

@&#8203;@&#8203; \-120\,11 \+120\,10 @&#8203;@&#8203;
 
 =item \*
 
\-Previous versions of File&#8203;::Find were guaranteed to call an I\<lstat>
\-before the user's C\<wanted\(\)> function was called\, but this is no
\-longer the case\.  Since this depends on C\<$File&#8203;::Find&#8203;::dont\_use\_nlink>\, $^O\,
\-and other factors\, fast file checks involving C\<\_> are not recommended
\-unless C\<wanted\(\)> calls I\<lstat> first\.
\+It is guaranteed that an I\<lstat> has been called before the user's
\+C\<wanted\(\)> function is called\. This enables fast file checks involving S\<\_>\.
\+Note that this guarantee no longer holds if I\<follow> or I\<follow\_fast>
\+are not set\.
 
 =item \*
 

==== //depot/perl/pod/perlfaq3\.pod\#89 \(text\) ====

@&#8203;@&#8203; \-85\,10 \+85\,12 @&#8203;@&#8203;
&nbsp;     use File&#8203;::Find;
&nbsp;     my @&#8203;files;
 
\-    find sub \{ push @&#8203;files\, $File&#8203;::Find&#8203;::name if \-f \_ && /\\\.pm$/ \}\,
\-         @&#8203;INC;
\+    find\(
\+      sub \{ push @&#8203;files\, $File&#8203;::Find&#8203;::name if \-f $File&#8203;::Find&#8203;::name && /\\\.pm$/ \}\,
\+      @&#8203;INC
\+    \);
 
\-       print join "\\n"\, @&#8203;files;
\+    print "$\_\\n" for @&#8203;files;
 
 If you simply need to quickly check to see if a module is
 available\, you can check for its documentation\.  If you can
